File: ios/Classes/UpdateManager.swift
//
//  UpdateManager.swift
//  mcumgr_flutter
//
//  Created by Mykola Kibysh on 11/12/2020.
//

import Foundation
import iOSMcuManagerLibrary
import CoreBluetooth

protocol UpdateManagerDelegate: AnyObject {
    func updateCompleted(manager: UpdateManager)
}

class UpdateManager {
    let transport: McuMgrBleTransport
    let progressStreamHandler: StreamHandler
    let stateStreamHandler: StreamHandler
    let logStreamHandler: StreamHandler
    let peripheral: CBPeripheral
    let updateLogger: UpdateLogger
    
    private (set) lazy var dfuManager: FirmwareUpgradeManager = FirmwareUpgradeManager(transporter: self.transport, delegate: self)
    
    init(peripheral: CBPeripheral, progressStreamHandler: StreamHandler, stateStreamHandler: StreamHandler, logStreamHandler: StreamHandler, updateLogger: UpdateLogger) {
        self.peripheral = peripheral
        self.transport = McuMgrBleTransport(peripheral)
        self.progressStreamHandler = progressStreamHandler
        self.stateStreamHandler = stateStreamHandler
        self.updateLogger = updateLogger
        self.logStreamHandler = logStreamHandler
    }
    
    func update(hash: Data, data: Data, config: FirmwareUpgradeConfiguration) throws {
        dfuManager.logDelegate = updateLogger
        try dfuManager.start(hash: hash, data: data, using: config)
    }
    
    func update(images: [ImageManager.Image], config: FirmwareUpgradeConfiguration) throws {
        dfuManager.logDelegate = updateLogger
        try dfuManager.start(images: images)
    }
    
    func pause() {
        if dfuManager.isInProgress() {
            dfuManager.pause()
        }
    }
    
    func resume() {
        if dfuManager.isPaused() {
            dfuManager.resume()
        }
    }

    func cancel() {
        dfuManager.cancel()
    }
}

extension UpdateManager: FirmwareUpgradeDelegate {
    func upgradeDidStart(controller: FirmwareUpgradeController) {
        
    }
    
    func upgradeStateDidChange(from previousState: FirmwareUpgradeState, to newState: FirmwareUpgradeState) {
        do {
            let oldProtoState = previousState.toProto()
            let newProtoState = newState.toProto()
            var changes = ProtoUpdateStateChanges()
            changes.oldState = oldProtoState
            changes.newState = newProtoState
            
            let arg = ProtoUpdateStateChangesStreamArg(updateStateChanges: changes, peripheral: peripheral)
            let data = try arg.serializedData()
            stateStreamHandler.sink?(FlutterStandardTypedData(bytes: data))
        } catch let e {
            let error = FlutterError(error: e, code: ErrorCode.flutterTypeError)
            stateStreamHandler.sink?(error)
        }
    }
    
    func upgradeDidComplete() {
        var stateChangesArg = ProtoUpdateStateChangesStreamArg(updateStateChanges: nil, peripheral: peripheral)
        stateChangesArg.done = true
        
        var progressArg = ProtoProgressUpdateStreamArg(progressUpdate: nil, peripheral: peripheral)
        progressArg.done = true
        
        var logArg = ProtoLogMessageStreamArg()
        logArg.uuid = peripheral.identifier.uuidString
        logArg.done = true
        
        do {
            let statusData = try stateChangesArg.serializedData()
            stateStreamHandler.sink?(FlutterStandardTypedData(bytes: statusData))
            
            let progressData = try progressArg.serializedData()
            progressStreamHandler.sink?(FlutterStandardTypedData(bytes: progressData))
            
            let logData = try logArg.serializedData()
            logStreamHandler.sink?(FlutterStandardTypedData(bytes: logData))
        } catch let e {
            let error = FlutterError(error: e, code: ErrorCode.flutterTypeError)
            stateStreamHandler.sink?(error)
        }
    }
    
    func upgradeDidFail(inState state: FirmwareUpgradeState, with error: Error) {
        var changes = ProtoUpdateStateChanges()
        changes.oldState = state.toProto()
        changes.newState = state.toProto()
        
        var arg = ProtoUpdateStateChangesStreamArg(updateStateChanges: changes, peripheral: peripheral)
        arg.error = ProtoError(localizedDescription: error.localizedDescription)
        
        var progressArg = ProtoProgressUpdateStreamArg(progressUpdate: nil, peripheral: peripheral)
        progressArg.done = true
        
        var logArg = ProtoLogMessageStreamArg()
        logArg.uuid = peripheral.identifier.uuidString
        logArg.done = true
        
        do {
            let data = try arg.serializedData()
            stateStreamHandler.sink?(FlutterStandardTypedData(bytes: data))
            
            let progressData = try progressArg.serializedData()
            progressStreamHandler.sink?(FlutterStandardTypedData(bytes: progressData))
            
            let logData = try logArg.serializedData()
            logStreamHandler.sink?(FlutterStandardTypedData(bytes: logData))
        } catch let e {
            let error = FlutterError(error: e, code: ErrorCode.flutterTypeError)
            stateStreamHandler.sink?(error)
        }
    }
    
    func upgradeDidCancel(state: FirmwareUpgradeState) {
        do {
            var changes = ProtoUpdateStateChanges()
            changes.canceled = true
            changes.oldState = state.toProto()
            changes.newState = state.toProto()
            
            let arg = ProtoUpdateStateChangesStreamArg(updateStateChanges: changes, peripheral: peripheral)
            stateStreamHandler.sink?(FlutterStandardTypedData(bytes: try arg.serializedData()))
        } catch let e {
            let error = FlutterError(error: e, code: ErrorCode.flutterTypeError)
            stateStreamHandler.sink?(error)
        }
    }
    
    func uploadProgressDidChange(bytesSent: Int, imageSize: Int, timestamp: Date) {
        do {
            var progressUpdate = ProtoProgressUpdate()
            progressUpdate.bytesSent = UInt64(Int64(bytesSent))
            progressUpdate.imageSize = UInt64(Int64(imageSize))
            progressUpdate.timestamp = timestamp.timeIntervalSince1970
            
            let arg = ProtoProgressUpdateStreamArg(progressUpdate: progressUpdate, peripheral: peripheral)
            let data = try arg.serializedData()
            let flutterData = FlutterStandardTypedData(bytes: data)
            progressStreamHandler.sink?(flutterData)
        } catch let e {
            let error = FlutterError(error: e, code: ErrorCode.flutterTypeError)
            progressStreamHandler.sink?(error)
        }
    }
}

File: android/src/main/kotlin/no/nordicsemi/android/mcumgr_flutter/UpdateManager.kt
package no.nordicsemi.android.mcumgr_flutter

import android.util.Log
import android.util.Pair
import io.runtime.mcumgr.ble.McuMgrBleTransport
import io.runtime.mcumgr.dfu.FirmwareUpgradeCallback
import io.runtime.mcumgr.dfu.FirmwareUpgradeController
import io.runtime.mcumgr.dfu.FirmwareUpgradeManager
import io.runtime.mcumgr.exception.McuMgrException
import no.nordicsemi.android.mcumgr_flutter.ext.shouldLog
import no.nordicsemi.android.mcumgr_flutter.ext.toProto
import no.nordicsemi.android.mcumgr_flutter.gen.*
import no.nordicsemi.android.mcumgr_flutter.logging.LoggableMcuMgrBleTransport
import no.nordicsemi.android.mcumgr_flutter.utils.StreamHandler
import java.security.MessageDigest


// Extensions of ByteArray to get sha1 hash
val ByteArray.sha1: String
	get() {
		val bytes = MessageDigest.getInstance("SHA-1").digest(this)
		return bytes.joinToString("") {
			"%02x".format(it)
		}
	}

data class FirmwareUpgradeConfiguration(
	val estimatedSwapTime: Long = 0,
	val eraseAppSettings: Boolean = true,
	val pipelineDepth: Int = 1,
	val byteAlignment: Int = 4,
	val reassemblyBufferSize: Long = 0
)

class UpdateManager(
		transport: McuMgrBleTransport,
		private val updateStateStreamHandler: StreamHandler,
		private val updateProgressStreamHandler: StreamHandler,
		private val logStreamHandler: StreamHandler
): FirmwareUpgradeCallback {
	private val manager: FirmwareUpgradeManager = FirmwareUpgradeManager(transport, this)
	private val address: String = transport.bluetoothDevice.address
	private val transport: LoggableMcuMgrBleTransport = transport as LoggableMcuMgrBleTransport

	init {
		manager.setMemoryAlignment(4)
		manager.setEstimatedSwapTime(5000)
		manager.setWindowUploadCapacity(3)
		manager.setMode(FirmwareUpgradeManager.Mode.CONFIRM_ONLY)
	}

	/**
	 * Start the upgrade.
	 *
	 * The specified image file will be sent to the target using the
	 * given transport, then verified using test command. If test successful, the reset
	 * command will be sent. The device should boot with the new firmware.
	 * The manager will try to connect to the SMP server on the new firmware and confirm
	 * the upload.
	 *
	 * @param firmware The firmware to be sent.
	 */

	private val TAG: String? = "MyActivity"

	fun start(images: List<Pair<Int, ByteArray>>, config: FirmwareUpgradeConfiguration?) {
		if (config != null) {
			manager.setMemoryAlignment(config.byteAlignment)
			manager.setEstimatedSwapTime(config.estimatedSwapTime.toInt())
			manager.setWindowUploadCapacity(config.pipelineDepth)
		}
		// print images to log
		images.forEach {
			val imageNumber = it.first
			val image = it.second
			// Get sha1 hash of image
			val sha1 = image.sha1

			// Log image and sha1
			Log.d(TAG, "Image $imageNumber: ${image.size} bytes, sha1: $sha1")
		}
		manager.start(images, config?.eraseAppSettings ?: true)
	}
	/** Pause the firmware upgrade. */
	fun pause() {
		if (!isPaused) {
			transport.setLoggingEnabled(true)
			manager.pause()
		}
	}
	/** Resume a paused firmware upgrade. */
	fun resume() {
		if (isPaused) {
			manager.resume()
			transport.setLoggingEnabled(false)
		}
	}
	/** Cancel the transfer. */
	fun cancel() = manager.cancel()
	/** True if the firmware upgrade is paused, false otherwise. */
	var isPaused = manager.isPaused
	/**	True if the firmware upgrade is in progress, false otherwise. */
	var isInProgress = manager.isInProgress
	/** Read all logs */
	fun readAllLogs() : ProtoLogMessageStreamArg {
		return (manager.transporter as? LoggableMcuMgrBleTransport)!!.readLogs()
	}

	override fun onUpgradeStarted(controller: FirmwareUpgradeController?) {
		transport.setLoggingEnabled(true)
	}

	override fun onStateChanged(prevState: FirmwareUpgradeManager.State?, newState: FirmwareUpgradeManager.State?) {
		transport.setLoggingEnabled(newState!!.shouldLog())
		transport.log("State changed: $prevState -> $newState")

		val changes = ProtoUpdateStateChanges(
			oldState = prevState!!.toProto(),
			newState = newState!!.toProto(),
		)
		val stateChangesArg = ProtoUpdateStateChangesStreamArg(
			uuid = address,
			updateStateChanges = changes,
		)
		updateStateStreamHandler.sink?.success(stateChangesArg.encode())
	}

	override fun onUpgradeCompleted() {
		transport.setLoggingEnabled(true)
		transport.log("Upgrade completed")

		val changes = ProtoUpdateStateChanges(
			newState = ProtoUpdateStateChanges.FirmwareUpgradeState.SUCCESS,
		)
		val successStateChanges = ProtoUpdateStateChangesStreamArg(
			uuid = address,
			updateStateChanges = changes,
		)
		updateStateStreamHandler.sink?.success(successStateChanges.encode())

		val stateChangesArg = ProtoUpdateStateChangesStreamArg(
			uuid = address,
			done = true,
		)
		updateStateStreamHandler.sink?.success(stateChangesArg.encode())

		val progressArg = ProtoProgressUpdateStreamArg(
			uuid = address,
			done = true,
		)
		updateProgressStreamHandler.sink?.success(progressArg.encode())

		val logArg = ProtoLogMessageStreamArg(
			uuid = address,
			done = true,
		)
		logStreamHandler.sink?.success(logArg.encode())
	}

	override fun onUpgradeFailed(state: FirmwareUpgradeManager.State?, error: McuMgrException?) {
		transport.setLoggingEnabled(true)
		transport.log("Upgrade failed")

		val changes = ProtoUpdateStateChanges(
			oldState = state!!.toProto(),
			newState = state.toProto(),
		)
		val stateChangesArg = ProtoUpdateStateChangesStreamArg(
			uuid = address,
			updateStateChanges = changes,
			error = ProtoError(
				localizedDescription = error?.message ?: "Unknown error",
			),
		)
		updateStateStreamHandler.sink?.success(stateChangesArg.encode())
	}

	override fun onUpgradeCanceled(state: FirmwareUpgradeManager.State?) {
		transport.setLoggingEnabled(true)
		transport.log("Upgrade canceled")

		val changes = ProtoUpdateStateChanges(
			oldState = state!!.toProto(),
			newState = state.toProto(),
			canceled = true,
		)
		val stateChangesArg = ProtoUpdateStateChangesStreamArg(
			uuid = address,
			updateStateChanges = changes,
		)
		updateStateStreamHandler.sink?.success(stateChangesArg.encode())
	}

	override fun onUploadProgressChanged(bytesSent: Int, imageSize: Int, timestamp: Long) {
		val progress = ProtoProgressUpdate(
			bytesSent = bytesSent.toLong(),
			imageSize = imageSize.toLong(),
			timestamp = timestamp.toDouble(), // convert to seconds
		)
		val arg = ProtoProgressUpdateStreamArg(
			uuid = address,
			progressUpdate = progress,
		)
		updateProgressStreamHandler.sink?.success(arg.encode())
	}
}